R/grade_this.R

```diff
@@ -269,6 +274,24 @@ NULL
 #' @export
 .user_code <- placeholder(".user_code")
 
+empty_code <- function(code, check_null = FALSE) {
+  if (check_null) {
+    empty_code <- is.null(code) || length(code) == 0 || !nzchar(code)
+  } else {
+    empty_code <- length(code) && !nzchar(code)
+  }
```

Comment:
Having a function just for checking for empty code lets you simply the logic using early returns.

```suggestion
  if (isTRUE(check_null) && is.null(code)) {
    return(TRUE)
  }
  
  length(code) == 0 || !nzchar(code)
```

or maybe the non-NULL empty code test should be


```suggestion
  if (isTRUE(check_null) && is.null(code)) {
    return(TRUE)
  }
  
  length(code) == 0 || all(!nzchar(trimws(code)))
```

Reply:
Since the `check_null` argument is only `TRUE` in one place, I just moved the logic out of the function altogether in [451763c](https://github.com/rstudio/gradethis/pull/288/commits/451763c4f9ff99ae55e906cb34de3b25066913be)

Reply:
Yeah and `length(NULL) == 0` so we'll catch it anyway.

`str_trim()` is a good idea, but we only officially support back to 3.4.

Reply:
Testing locally, I changed `is_empty_code()` to

```r
is_empty_code <- function(code) {
  length(code) == 0 || all(!nzchar(str_trim(code)))
}
```

and no tests failed. Reading between the lines, I think that when we were including this test in `grade_this()` it would have made sense to allow `.user_code` to be `NULL` since we might not want to be overly restrictive. But now that we're applying the test in a more targeted way for just `grade_this_code()` and `fail_if_code_feedback()`, it's okay for `is_empty_code()` to be a little more general.

Reply:
I made the change above in [2fce7b5](https://github.com/rstudio/gradethis/pull/288/commits/2fce7b57cfe7f417714ee9f39168d453b1ea70d4) and then moved `grade_code_is_empty()` to `grade_this_code.R` in [61a2919](https://github.com/rstudio/gradethis/pull/288/commits/61a2919e78476e3622992d1e60c6de7fb4481317). I'm going to merge this to unblock other PRs but if I've misunderstood something we can fix it in a new PR.

Reply:
> I think that when we were including this test in `grade_this()` it would have made sense to allow `.user_code` to be `NULL`

Yeah, when the check happened in `grade_this()`, it caused a lot of test failures because a lot of the tests for `grade_this()` didn't define `.user_code`. Now that `is_empty_code()` is only called when code checking is going to happen, I think it makes sense to check if the code is an empty string _or_ `NULL`.